server/server.go

```diff
@@ -42,72 +55,362 @@ func HandleWebsocketConnection(serverHTTPResponse http.ResponseWriter, clientHTT
+	if websocketConnectionrequestAuthErr != nil {
 		// If the authentication failed, terminate the websocket connection to the client
-		log.Println("Authentication Failed:", websocketConnectionStateErr)
-		websocketConnection.WriteMessage(websocket.BinaryMessage, []byte(websocketConnectionStateErr.Error()))
+		log.Println("Authentication Failed:", websocketConnectionrequestAuthErr)
+		websocketConnection.WriteMessage(websocket.BinaryMessage, []byte(websocketConnectionrequestAuthErr.Error()))
 		log.Println("Terminating websocket connection to client.")
+		// Close the websocket connection
+		websocketConnection.Close()
 		return
 	}
+	// Check if there are pending messages to be delivered when the client comes back online
+	if messagesToBeDelivered, ok := multiUserChatMessage[authenticatedIdentityPublicKeyHexClient]; ok {
+		deliverMessages(websocketConnection, messagesToBeDelivered)
+	} // if messagesToBeDelivered
 
-	// Notify the client that authentication succeeded
-	websocketConnection.WriteMessage(websocket.BinaryMessage, []byte("Authentication Succeeded"))
-
+	websocketConnectionProcessMessageErr := processMessage(websocketConnection, authenticatedIdentityPublicKeyHexClient)
+	for websocketConnectionProcessMessageErr == nil {
+		websocketConnectionProcessMessageErr = processMessage(websocketConnection, authenticatedIdentityPublicKeyHexClient)
+	}
+	// Close the websocket connection
+	websocketConnection.Close()
+	return
+	// Read first message from client
 } // func handleWebsocketConnection
 
-func requireAuth(websocketConnection *websocket.Conn) (int, error) {
-	// Create a variable to keep track of the websocket connection state
-	// Create a byte sequence that we would ask the client to sign
-	byteSequenceToSign := []byte("hi")
-	// Initialize the Auth protobuf structure
-	protobufAuth := backendProtobuf.Auth{}
-	// Fill in the byte sequence that we would ask the client to sign
-	protobufAuth.ToSign = byteSequenceToSign
+func processMessage(websocketConnection *websocket.Conn, authenticatedIdentityPublicKeyHexClient string) error {
+	// Initialize an empty variable to hold the protobuf message
+	var messageFromClientProtobuf backendProtobuf.BackendMessage
+	// Read a message from a client over the websocket connection
+	_, messageFromClientBytes, readMessageErr := websocketConnection.ReadMessage()
+	if readMessageErr != nil {
+		// If there is an error while reading the response from the client, return the error
+		return readMessageErr
+	}
+	// Unmarshal the protobuf bytes from the message we received into our protobuf message structure
+	protoUnmarshalErr := proto.Unmarshal(messageFromClientBytes, &messageFromClientProtobuf)
+	if protoUnmarshalErr != nil {
+		// If there is an error while unmarshaling the response from the client, return the error
+		return protoUnmarshalErr
+	}
+	// Outer master switch, cases are valid if they are true
+	switch {
```

Comment:
maybe add a case for a missing request id

Reply:
Yes, request id should be a required parameter in a valid message.